### ISSUE TITLE
API: tags on media

### ISSUE THREAD

**Issue by fernao** [RedeMocambos/baobaxia#89]
Implement tags on media, using git annex metadata

**Comment by fernao** [comment]
For gallery #104, we've reactivated tags on media/serializer.py, in order to have a way to aggregate medias. So, our concept of gallery is a simple tag added to the content, but works only locally (based on django tags).
Any news about tags implementation at git annex level?

**Comment by agger-magenta** [comment]
I've been thinking about this today, and I've now decided that I think I know how it should be done. I'm looking at it right now & will let you know in the next few days.

**Comment by agger-magenta** [comment]
There are two directions from which tags can arrive on the media.

One is when the user saves a media file and supplies some tags. As these tags originate from the local mucua, they must be supplied as metadata tags on the media, if not present already.

Another is when we synchronize a file from git annex. In this case we receive a file which has some git-annex metadata attached. If the file is new, the tags will all be from different namespaces, but if it's not new, there may be some local ones as well. Right now it makes sense for me to distinguish between these two situations. In the code, we may wish to handle them the same place, e.g. in the media's post save signal handler - or not.

From a git annex metadata perspective, all tags will have a namespace, and the namespace will be the name of the mucua where it originated (by being set on some media). For backwards compatibility, we can assume tags with no namespace to always be equivalent to "local mucua". A tag, e.g.  "permacultura", e.g. on the mucua "dpadua" with no namespace information will thus be used to tag the media file with "dpadua:permacultura".

**Comment by agger-magenta** [comment]
The above considerations mean that it'll probably not be the best idea to do the tag synchronization on the post-save signal.

When synchronizing from git-annex, it should be done in the `create_objects_from_file` function in `media/serializers.py`. Upon create, update etc. it should be done at a point where we know that the media as well as the tags are securely saved.

**Comment by agger-magenta** [comment]
Question: How are tags supposed to come into existence?

Right now, you can only add tags that already exist as instances of the Tag Django model. Thus, the only way to create tags are through Django admin and by tagging with git-annex metadata (once that works). This makes perfect sense at least in the way that spelling mistakes are avoided and thus we don't end up with several versions of the same tag, but it doesn't make too much sense when you upload a gallery and want to name it with some presumably new tag.

How is this supposed to end up working?

**Comment by agger-magenta** [comment]
It works! 

I've tested it in the sense that I've created a document, added one tag and checked that it was tagged with git-annex metadata. 

I've also tagged the same document from git-annex, called create_objects_from_files and seen that a new tag was created and the document tagged with it in Django.

In the next couple of days I expect to refactor the synchronization code to avoid DRY violations and make it somewhat clearer - and test some edge cases.
